### ISSUE TITLE
Start using "Best available" location in GrantNav location filters

### ISSUE THREAD

**Issue by mariongalley** [ThreeSixtyGiving/grantnav#867]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. E.g. I'm always frustrated when [...]

GrantNav currently does not use Beneficiary location in the location filters. This causes the two following issues:
- Grants from Funders who provide Beneficiary location but not Recipient location do not appear in location-based searches at all, despite those funders providing good quality geodata according to our guidance.
- Grants from Funders who provide both Beneficiary location and Recipient location appear in location-based searches under their Recipient location, which can exacerbate the "head office effect".

**Describe the solution you'd like**

Start using "Best available" location in GrantNav location filters i.e. defaults to Beneficiary and falls back on Recipient location (does not use derived Recipient location).

This means the appearance of the location filters remains exactly the same, but when you select a location, the results from that search show grants where the Beneficiary location corresponds to the selected value(s), and for grants with no Beneficiary location, the results show grants where the Recipient location corresponds to the selected value(s).

Grants with neither Beneficiary nor Recipient location data will continue to not be returned in geographic searches.

**Describe alternatives you've considered**

We considered duplicating the location filters and having one set that searches based on Recipient location and one that searches based on Beneficiary location. We also considered having a single location filter that defaults to Best Available location, but also allows the user to switch to searching just within Recipient location or just within Beneficiary location using a toggle. We rejected both for these reasons:

- Most GrantNav users do not understand the difference between the two and would not know what to do when presented with this choice
- Beneficiary location is not widely used enough for a search on Beneficiary location to return meaningful results

**Additional context**

**Comment by KDuerden** [comment]
> We rejected both for these reasons:
> Most GrantNav users do not understand the difference between the two and would not know what to do when presented with this choice
> Beneficiary location is not widely used enough for a search on Beneficiary location to return meaningful results

My understanding was that we were prioritising getting Beneficiary location to work in GrantNav by using best available, but that we would as a second phase allow for the toggling between beneficiary and recipient location. I wouldn't be comfortable making this change without the plan to introduce the option because I think it will make the location features less useful overall.

Allowing for recipient or beneficiary location to be filtered (via a toggle), in addition to "best available" means users can still find organisations are based in their area.

For many publishers sharing beneficiary location data, this is at a wider region or country level. Removing the option to also show their data at district level when the recipient location data is provided, would make data that normally appears in filters disappear, removing one of the useful ways they are used to viewing their data.

This doesn't change the purpose of this issue, but I want to note my different understanding of the longer term goal.

**Comment by mariongalley** [comment]
> My understanding was that we were prioritising getting Beneficiary location to work in GrantNav by using best available, but that we would as a second phase allow for the toggling between beneficiary and recipient location. 

Yes sorry, I've checked my notes from our discussion and can confirm that's the case. This is really useful to state up front in case this has any implications for the architecture of the solution, so thanks!

**Comment by michaelwood** [comment]
Just to confirm - Does "filter" in this issue mean the filters on the left hand side or a [search for a location](https://grantnav.threesixtygiving.org/search?query=London&default_field=additional_data.recipientLocation&sort=_score+desc)?  

Filters on the left hand side have to act on a single field in the data so any ordering of preferred location source would need to be added as additional data field in the grant data in the datastore.

**Comment by KDuerden** [comment]
> Filters on the left hand side have to act on a single field in the data so any ordering of preferred location source would need to be added as additional data field in the grant data in the datastore.

Like an equivalent of the canonical name / id?

**Comment by mariongalley** [comment]
> Just to confirm - Does "filter" in this issue mean the filters on the left hand side or a search for a location?

Good question! Yeah I'd expect the filters on the left hand side to have the priority order Beneficiary Location then Recipient Location. For the main search bar filter, would we want it to include any locations @KDuerden ?

**Comment by KDuerden** [comment]
>For the main search bar filter, would we want it to include any locations?

I think we need to have the filter and location search behave in consistent ways, otherwise explaining to users how and why their results are different could be challenging.

However I'm not sure how this fits with allowing for toggling between best, recipient and beneficiary at a future point.

**Comment by mariongalley** [comment]
To avoid triplicating the Locations filter on the top search bar, we might have to call it Locations (Best Available), and have it only use Best Available location.

**Comment by mariongalley** [comment]
> Maybe in the Search bar that could be renamed to "All Location fields" or just "All locations" to indicate that we will search across multiple fields for the entered term.

Agreed this means it will search for the term entered across all locations, not just in the best available location

**Comment by michaelwood** [comment]
Given the discussions we recently had, do we want to have fully separate filters for the different location types? e.g. Mock up -

![image](https://user-images.githubusercontent.com/493697/204522850-c83b3edf-4934-4970-be6a-537b97e07850.png)

**Comment by mariongalley** [comment]
@michaelwood For now as we have limited time and budget we're happy to just have the location filter default to Best Available, and then we'll add the option to choose between different types of location later.

When we do develop this though, I think it's better to have one set of Location filters and ask people to choose between Best Available, Beneficiary or Recipient Location. This is for a few reasons:
1. There's limited space on the left and we'd have to triple the amount Location filters, which would take up too much space.
2. Defaulting to Best Available means people aren't immediately confronted with a choice about *which* location filter to use, which for many people will be confusing.
3. While it may be interesting to search for grants with a Recipient in one place and a Beneficiary in another, in most cases there won't be much or any data. It would also be hard to define what the filters should do if someone selects a location in both the Best Available filter and the Beneficiary or Recipient filters.
4. Another Benefit of the multi-way toggle is that we can add the option to choose No Location in the future, so there's a 4-way toggle between Best Available/Beneficiary/Recipient/No location and if you pick the latter you can't define a location, but you get to see all the grants that were excluded from your other searches.

I think we'd need to design a new component for a multi-way toggle - something a bit like this above the location filters:
![image](https://user-images.githubusercontent.com/82958988/204528298-669ee2a6-6ba0-4c0e-8078-5f0457a979b6.png)

**Comment by michaelwood** [comment]
Okay. No problem. Will leave as is, possibly some docs in the grantnav help will be able to clarify what "best location" means.

**Comment by mariongalley** [comment]
Yeah we'll need to update https://help.grantnav.threesixtygiving.org/en/latest/locations.html as it will be inaccurate with this update - I'll start a doc for the copy update for this then

**Comment by michaelwood** [comment]
Think we're getting there on this: http://grantnav.360dev1.vs.mythic-beasts.com/search?query=%2A&default_field=%2A&sort=_score+desc&fundingOrganization=GB-CHC-1080418&recipientRegionName=South+West

I have some slight adjustment to the rules as I think using locationLookup source = beneficiaryLocation turns out to be more reliable than the other geo methods in the datastore so I'm going to allow that data to overwrite data that has been set by the previous geo source of local file geo codelists.

If we had a bit more time some refactoring of the multiple methods of geo location would be a good idea.

**Comment by mariongalley** [comment]
> If we had a bit more time some refactoring of the multiple methods of geo location would be a good idea.

How much more time are we talking? Happy for this to be raised as a separate ticket to be done after Christmas potentially

**Comment by michaelwood** [comment]
tldr: about a week

Currently grantnav uses fields provided by an old additional_data mechanism that existed from about 5 years ago (where it added the data directly to each Grant on import into GrantNav - this was before the datastore existed). 

Insights uses the newer locationLookup fields. For now I can map locationLookup fields into the old mechanism, when available but really we could do with deprecating it fully and retiring that module so that both insights and grantnav use exactly the same mechanism. We can do that with some refactoring but it needs to be done pretty carefully as it effects every places in GrantNav that we do anything 'geo' with.

The testing cycles of datastore development <-> GrantNav are quite slow as each time the additional_data mechanism is updated it takes an hour to regenerate the data. Then the data has to be reloaded into grantnav. We can work with a smaller dataset, but having done this before we have missed some of the corner cases.

**Comment by mariongalley** [comment]
@michaelwood Thank you, that sounds good - could you please raise a separate ticket for that work and include the ~ week estimate on it so we can add it to next year's contact? I think anything more than 4-5 days will want its own grant line for next year to avoid the iterative improvements being a catch-all. Then we can add it to the project board for 'within a few months' probably

**Comment by michaelwood** [comment]
Believe we have fixed this issue. our mechanism now tries beneficiaryLocation and then recipientLocation as the best location for the impact of the grant. We have some legacy bits in Grantnav to tidy up at some point, these have been commented in the code.

**Comment by mariongalley** [comment]
@michaelwood Please can you raise the separate ticket for refactoring multiple methods of geolocation?

**Comment by michaelwood** [comment]
> @michaelwood Please can you raise the separate ticket for refactoring multiple methods of geolocation?

Done - https://github.com/ThreeSixtyGiving/grantnav/issues/953
